R/utils.R

```diff
@@ -504,3 +504,175 @@ download_with_retry <- function(...) {
   }
   return(rc)
 }
+
+#' Checks if NONMEM run is done by looking for "Stop Time" in .lst file
+#'
+#' Returns `TRUE` if the model appears to be finished running and `FALSE` otherwise.
+#' @param mod either a bbi_nonmem_model object or a path to an .lst file
+#' @export
+check_nonmem_finished <- function(mod) {
+  UseMethod("check_nonmem_finished")
+}
+
+
+#' Check if NONMEM is finished (character)
+#' @param mod character string of NONMEM model
+#'
+#' @importFrom checkmate assert_string
+#' @importFrom readr read_lines
+#'
+#' @describeIn check_nonmem_finished takes a file path (character string) of `NONMEM` `.lst` file.
+#' @export
+check_nonmem_finished.character <- function(mod) {
+  assert_string(mod)
+
+  # look for model to be finished and then test output
+  model_finished <- try({suppressWarnings({
+    read_lines(mod) %>%
+      str_detect("Stop Time") %>%
+      any()
+  })})
+  return(isTRUE(model_finished))
+}
+
+#' Check if NONMEM is finished (bbi model)
+#' @param mod bbi model object
+#'
+#' @describeIn check_nonmem_finished takes a `bbi_nonmem_model` object.
+#' @export
+check_nonmem_finished.bbi_nonmem_model <- function(mod) {
+  # check that output dir was created (should happen as soon as model is submitted)
+  if (!fs::dir_exists(get_output_dir(mod, .check_exists = FALSE))) {
+    return(TRUE) # if missing then this failed right away, likely for some bbi reason
+  }
+
+  # pass .lst path to character dispatch
+  build_path_from_model(mod, ".lst") %>%
+    check_nonmem_finished()
+}
+
+#' Wait for NONMEM models to finish
+#' @param mod a `bbi_nonmem_model` object, or list of `bbi_nonmem_model` objects.
+#' @param time_limit integer for maximum number of seconds in total to wait before continuing (exiting function)
```

Comment:
Maybe add something to clarify "even if the run does not appear to have finished"